[JSC] Align `Intl.DateTimeFormat` with ECMA-402

Gumichocopengin8 opened the pull request:
#### 0c97eb8491636a76c41018c0ab6bf17ad973af8d
<pre>
[JSC] Align `Intl.DateTimeFormat` with ECMA-402
<a href="https://bugs.webkit.org/show_bug.cgi?id=303023">https://bugs.webkit.org/show_bug.cgi?id=303023</a>

Reviewed by NOBODY (OOPS!).

This patch aligns `Intl.DateTimeFormat` with ECMA-402
by adding fallback for calendar option related to `islamic` [1].

[1]: <a href="https://tc39.es/proposal-intl-era-monthcode/#sec-createdatetimeformat">https://tc39.es/proposal-intl-era-monthcode/#sec-createdatetimeformat</a>

* JSTests/stress/intl-datetimeformat.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::isFallbackIslamic):
(JSC::intlAvailableCalendars):
* Source/JavaScriptCore/runtime/IntlObject.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c97eb8491636a76c41018c0ab6bf17ad973af8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141320 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85804 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6118 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102308 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69663 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136692 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119897 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83111 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4668 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2279 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125822 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38013 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143967 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132259 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5924 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6008 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5065 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110878 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4521 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116152 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59671 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5977 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34450 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165222 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5823 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69441 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43164 "Found 3 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/regexp-zero-length-alternatives.js.layout, stress/regexp-named-capture-groups.js.dfg-eager, stress/simple-regexp-test-folding.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6069 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5931 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/IntlObject.cpp:
```diff
@@ -1531,6 +1531,12 @@ bool isWellFormedCurrencyCode(StringView currency)
+// #9 in https://tc39.es/proposal-intl-era-monthcode/#sec-createdatetimeformat
+bool isFallbackIslamic(const String& calendar)
+{
+    return calendar.endsWith("islamic"_s) || calendar.endsWith("islamic-rgsa"_s);
```
Step 9 sounds like it would be `==`, not `endsWith`. What’s the rationale for `endsWith`? Related, should we include a leading `-` in the suffixes we are checking for, like `"-islamic"_s`? Not sure I see enough test cases to clarify if this is OK.

Gumichocopengin8 replied:
I previously used `endsWith` because I thought some test cases didn't pass, but I overlooked something. All test cases pass with `==`, so I fixed it with `==`. 
Based on https://tc39.es/proposal-intl-era-monthcode/#sup-availablecalendars, I think we don't need a leading `-` to check because `en-u-ca-islamic`, `en-u-ca-islamic-tbla` and others are not listed.

darinadler commented on Source/JavaScriptCore/runtime/IntlObject.h:
```diff
@@ -159,6 +159,8 @@ String canonicalizeUnicodeLocaleID(const CString& languageTag);
 
 bool isWellFormedCurrencyCode(StringView);
 
+bool isFallbackIslamic(const String&);
```
```suggestion
bool isFallbackIslamic(StringView);
```

Gumichocopengin8 replied:
Fixed.

Gumichocopengin8 committed:
[JSC] Align `Intl.DateTimeFormat` with ECMA-402
https://bugs.webkit.org/show_bug.cgi?id=303023

Reviewed by NOBODY (OOPS!).

This patch aligns `Intl.DateTimeFormat` with ECMA-402
by adding fallback for calendar option related to `islamic` [1].

[1]: https://tc39.es/proposal-intl-era-monthcode/#sec-createdatetimeformat

* JSTests/stress/intl-datetimeformat.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::isFallbackIslamic):
(JSC::intlAvailableCalendars):
* Source/JavaScriptCore/runtime/IntlObject.h:

darinadler commented on Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
```diff
@@ -717,7 +717,7 @@ void IntlDateTimeFormat::initializeDateTimeFormat(JSGlobalObject* globalObject,
 
         m_calendar = WTFMove(calendar);
         // Handling "islamicc" candidate for backward compatibility.
-        if (m_calendar == "islamicc"_s)
+        if (m_calendar == "islamicc"_s || isFallbackIslamic(m_calendar))
             m_calendar = "islamic-civil"_s;
```
Must admit I don't understand why this transformation is necessary, and the existing comment certainly doesn't explain it.

darinadler commented on Source/JavaScriptCore/runtime/IntlObject.h:
```diff
@@ -159,6 +159,8 @@ String canonicalizeUnicodeLocaleID(const CString& languageTag);
 
 bool isWellFormedCurrencyCode(StringView);
 
+bool isFallbackIslamic(StringView);
```
I don't think this function name is ideal. It doesn't match any wording in the specification and it also doesn't make it clear what kind of string the argument is.

darinadler commented on Source/JavaScriptCore/runtime/IntlObject.cpp:
```diff
@@ -1531,6 +1531,12 @@ bool isWellFormedCurrencyCode(StringView currency)
     return currency.length() == 3 && currency.containsOnly<isASCIIAlpha>();
 }
 
+// #9 in https://tc39.es/proposal-intl-era-monthcode/#sec-createdatetimeformat
```
The specification says that something should be logged when this is used but I don't see that in the code we are adding.

darinadler reviewed (approved):
Seems OK, some further comments.

sosukesuzuki reviewed (commented):
[intl-era-monthcode](https://github.com/tc39/proposal-intl-era-monthcode) is still a stage 2 proposal. Therefore, I don't think this fix needs to be implemented yet. What do you think?

I think we should make `test262-runner` ignore `Intl.Era-monthcode` feature flag
